### PR TITLE
Gracefully handle Git deletions again

### PR DESCRIPTION
**Pull request by vemv**
## Brief

`impl/absolutize`, particularly with its `.exists` assertion,  had broken our ability to gracefully handle deleted files (staged or not).

This PR fixes that, adding an extensive integration test.

## QA plan

Test all cases in a real project (suggestion: a monorepo):

* Deleted file (staged for deletion or not)
* plain `core/format!`, `format-and-lint-project!`, `format-and-lint-branch!`

That's 6 scenarios - please let's test them all (I have in formatting-stack itself).

## Author checklist

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [x] Robustness (red paths, failure handling etc)
  * [x] Modular design
  * [x] Test coverage
  * [x] Spec coverage
  * [x] Documentation
  * [x] Security
  * [x] Performance
    * `defn all-files` peforms 3 git ops but they're fairly cheap (5-15ms each)
  * [x] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [ ] I have checked out this branch and reviewed it locally, running it
* [ ] I have QAed the functionality
* [ ] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Modular design
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

File: src/formatting_stack/strategies.clj [hunk 1]
```diff
@@ -25,25 +28,37 @@
...
+                       (set))
+          tracked (->> (impl/file-entries git-command "ls-files")
+                       (impl/absolutize git-command)
+                       (remove deleted))
+          untracked (->> (impl/file-entries git-command "ls-files" "--others" "--exclude-standard")
+                         (impl/absolutize git-command)
+                         (remove deleted))]
+      ;; Second `binding`, to ensure correct results
+      (binding [impl/*skip-existing-files-check?* false]
+        (speced/let [^::impl/existing-files corpus (into tracked untracked)]
```
Comment (thumbnail) [inline src/formatting_stack/strategies.clj]: you could remove some duplication by doing this:

```suggestion
          tracked   (impl/file-entries git-command "ls-files")
          untracked (impl/file-entries git-command "ls-files" "--others" "--exclude-standard")]
      ;; Second `binding`, to ensure correct results
      (binding [impl/*skip-existing-files-check?* false]
        (speced/let [^::impl/existing-files corpus (->> (into tracked untracked)
                                                        (impl/absolutize git-command)
                                                        (remove deleted))]
```

File: src/formatting_stack/strategies.clj [hunk 2]
```diff
@@ -64,9 +80,12 @@
   The diff is compared against the `:target-branch` option."
   [& {:keys [target-branch impl files blacklist]
       :or   {target-branch "master"
-             impl          (impl/file-entries git-command "diff" "--name-only" target-branch)
+             impl          (impl/file-entries git-command "diff" "--name-status" target-branch)
              blacklist     (git-not-completely-staged :files [])}}]
   (->> impl
+       (remove (fn [s]
+                 (-> s (string/starts-with? "D\t"))))
```
Comment (thumbnail) [inline src/formatting_stack/strategies.clj]: We should use `git-status/deleted-file?` here

File: src/formatting_stack/strategies.clj [hunk 3]
```diff
@@ -64,9 +80,12 @@
...
+       (map #(string/replace-first % #"(A|C|D|M|R|T|U|X|B)\t" "")) ;; See: `git diff --help`, `diff-filter` section
```
Comment (thumbnail) [inline src/formatting_stack/strategies.clj]: What about removing the first _word_ from the output? it seems less brittle

File: test/integration/formatting_stack/strategies.clj [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+(def ^File deletable-file (File. "test-resources/deletable.clj"))
+
+(def ^String createable-filename "test-resources/createable.clj")
+
+(def creatable-contents (pr-str '(ns foo)))
+
+(assert (-> deletable-file .exists))
+
+(assert (not (-> createable-filename File. .exists)))
+
```
Comment (thumbnail) [inline test/integration/formatting_stack/strategies.clj]: These should be inside of the test. The files don't have to exist/not-exists when _loading_ this ns, they should when i run the tests

File: test/integration/formatting_stack/strategies.clj [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        (sh "git" "checkout" (str deletable-file))
+        (assert (-> deletable-file .exists))))))
+
+(deftest git-diff-against-default-branch
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of deleted (but not staged) files"
+    (let [;; f-stack's first ever commit. This ensures a large, diverse corpus:
+          first-commit "ebb0ca8"]
```
Comment (thumbnail) [inline test/integration/formatting_stack/strategies.clj]: could be [`git rev-list --max-parents=0 HEAD`](https://stackoverflow.com/questions/5188914/how-to-show-first-commit-by-git-log). to keep the actual commit hash out of the source

Comment (thumbnail) [reply test/integration/formatting_stack/strategies.clj]: The let can be one level higher, (or a delayed def with the command above) preventing duplication on L77

File: test/integration/formatting_stack/strategies.clj [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      (testing "Added files don't show up"
+        (try
+          (sh "git" "add" "-A")
+          (is (= []
+                 (sut/git-not-completely-staged :files [])))
+          (finally
+            (sh "git" "reset" "."))))
+
+      (finally
+        (-> createable-filename File. .delete))))
```
Comment (thumbnail) [inline test/integration/formatting_stack/strategies.clj]: exceptions will be caught inside of the `is`, right? does this work how the way its intended?

File: src/formatting_stack/strategies.clj (hunk 1 above)
Comment (vemv) [reply src/formatting_stack/strategies.clj]: > you could remove some duplication by doing this:

Will try

File: src/formatting_stack/strategies.clj (hunk 2 above)
Comment (vemv) [reply src/formatting_stack/strategies.clj]: > We should use git-status/deleted-file? here

Nope, note `git-status`' ns docstring

File: src/formatting_stack/strategies.clj (hunk 3 above)
Comment (vemv) [reply src/formatting_stack/strategies.clj]: > What about removing the first word from the output? i

Dunno, removing arbitrary stuff can also be brittle.

unit-testing the `replace-first` can be a middle ground

File: test/integration/formatting_stack/strategies.clj (hunk 4 above)
Comment (vemv) [reply test/integration/formatting_stack/strategies.clj]: > These should be inside of the test.

Note that 2 deftests depend on this assertion. Having 2 deftests fail seems less clear, concise and immediate than having `(refresh)` fail as soon as an issue arises.

File: test/integration/formatting_stack/strategies.clj (hunk 5 above)
Comment (vemv) [reply test/integration/formatting_stack/strategies.clj]: > git rev-list --max-parents=0 HEAD.

Will try

File: test/integration/formatting_stack/strategies.clj (hunk 6 above)
Comment (vemv) [reply test/integration/formatting_stack/strategies.clj]: > exceptions will be caught inside of the is, right?

`finally` runs 100% unconditionally, whether an exception was thrown or not.

File: test/integration/formatting_stack/strategies.clj (hunk 4 above)
Comment (thumbnail) [reply test/integration/formatting_stack/strategies.clj]: Except that now i can refresh, delete the file and run the test. yielding unclear failures as well.

File: src/formatting_stack/strategies.clj (hunk 2 above)
Comment (thumbnail) [reply src/formatting_stack/strategies.clj]: ouch, okay :P

File: test/integration/formatting_stack/strategies.clj (hunk 4 above)
Comment (vemv) [reply test/integration/formatting_stack/strategies.clj]: > Except that now i can refresh, delete the file and run the test.

Running tests without a prior `(refresh)` seems just wrong. Our dev.clj file typically suggests to `(refresh)` prior to running any test

File: src/formatting_stack/strategies.clj (hunk 1 above)
Comment (vemv) [reply src/formatting_stack/strategies.clj]: Cannot be done: the two `(impl/absolutize git-command)` calls must be called within the former binding.

Thankfully, `integration.formatting-stack.strategies` fails if we applied this patch

**Comment by vemv** [comment]
Ready again!

**Comment by vemv** [comment]
👀!

Did you get exceptions? Supposedly there's a spec that would assert `impl/absolutize` output would correspond to an existing file

**Comment by thumbnail** [comment]
> Supposedly there's a spec that would assert impl/absolutize output would correspond to an existing file

It's probably happening when `*skip-existing-files-check?*` is bound to `true`

**Comment by vemv** [comment]
> Did you get exceptions?

(For clarity: at least the spec checking part was working, w/ nice Expound reports)

**Comment by thumbnail** [comment]
> Did you get exceptions?

Only the `ExceptionInfo`-exceptions with the validation errors.

<details>

```
-- Relevant specs -------

:formatting-stack.strategies.impl/existing-files:
  (clojure.spec.alpha/coll-of
   (nedap.speced.def/fn
    [s]
    (if
     formatting-stack.strategies.impl/*skip-existing-files-check?*
     true
     (clojure.core/-> s File. .exists))))

-- Spec failed --------------------

  (...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   "<SNIP>"
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...
   ...)

should satisfy

  (speced/fn
   [s]
   (if *skip-existing-files-check?* true (-> s File. .exists)))

-- Relevant specs -------

:formatting-stack.strategies.impl/existing-files:
  (clojure.spec.alpha/coll-of
   (nedap.speced.def/fn
    [s]
    (if
     formatting-stack.strategies.impl/*skip-existing-files-check?*
     true
     (clojure.core/-> s File. .exists))))

-------------------------
Detected 27 errors

Execution error (ExceptionInfo) at formatting-stack.strategies/all-files$fn (strategies.clj:47).
Validation failed
```

</details>

**Comment by vemv** [comment]
Some notes:

* `git status porcelain` always emits paths relative to the git root, not to the cwd
* `git diff --name-status` always emits paths relative to the git root, not to the cwd
* `git ls-files`, at least by default emits paths relative to the cwd

The latter, beng different, is a source of issues.

**Comment by vemv** [comment]
Kindly QA again. I'll be also, over a non-monorepo

**Comment by thumbnail** [comment]
staged/unstaged deleted files within the same module work as expected. However `strategies/all-files` only returns the files present in `(System/getProperty "user.dir")` (e.g. the module itself). Not the files outside of the module.

**Comment by vemv** [comment]
I thought about that.

It seems bit of a poor default to return files outside of the monorepo. Most folks will _not_ want files outside of the current project to be returned.

Else they will be formatting/linting files outside of what Clojure can `require`, so bugs would arise.

That's a non-issue for us `:monorepo` profile, or people doing similar stuff (https://clubhouse.io/blog/monolith-meet-mono-repo/). But it seems unsafe to assume that everyone will have that set up.

(Even among us :monorepo is optional)

**Comment by thumbnail** [comment]
> Else they will be formatting/linting files outside of what Clojure can require, so bugs would arise.

This is not necessarily true, the files might be on the classpath, but not in the current project. It might be specific to our monorepo setup, but running `project-formatter/format-and-lint-project!` for each project could make sense 🤔

**Comment by vemv** [comment]
> This is not necessarily true, the files might be on the classpath, but not in the current project.

Yes, but we cannot know with all certainty.

(Monorepo != "all projects can reach each other's set of namespaces")

Created https://github.com/nedap/formatting-stack/issues/139 + https://github.com/nedap/formatting-stack/issues/140 btw, which can be attended promptly but would not consider a blocker (as f-s is somewhat broken rn)

**Pull request merged**